File: go/tests/pressure/bmp384_test/main.go
//go:build linux && !tinygo

package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/pressure"
	"github.com/tuhde/Periph/go/periph/connection"
)

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

func main() {
	bus, _ := strconv.Atoi(envOr("I2C_BUS", "1"))
	addr64, _ := strconv.ParseUint(envOr("I2C_ADDR", "0x76"), 0, 8)

	tr, err := connection.NewI2CConnection(bus, uint8(addr64), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer tr.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	chip, err := pressure.NewBMP384Minimal(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}
	check("default_osr_p_4", chip.OsrP == 4)
	check("default_osr_t_1", chip.OsrT == 1)
	check("default_iir_2",   chip.Iir  == 2)

	t, err := chip.Temperature()
	check("temperature_in_range", err == nil && t >= -40 && t <= 85)
	p, err := chip.Pressure()
	check("pressure_in_range",    err == nil && p >= 300 && p <= 1250)

	full, err := pressure.NewBMP384Full(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new full:", err)
		os.Exit(2)
	}
	if _, err := full.IsDataReady(); err == nil {
		check("is_data_ready", true)
	} else {
		check("is_data_ready", false)
	}
	if err := full.Configure(2, 1, 1, 0x04); err == nil {
		check("configure_writes_through",
			full.OsrP == 2 && full.Iir == 1 && full.Odr == 0x04)
	} else {
		check("configure_writes_through", false)
	}
	if frames, err := full.FIFORead(); err == nil {
		check("fifo_read_returns_slice", frames != nil)
	} else {
		check("fifo_read_returns_slice", false)
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

File: rust/tests/pressure/bmp384_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Bmp384Full, Bmp384Minimal, MODE_FORCED, MODE_NORMAL};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x76);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp = Bmp384Minimal::new(dev, addr, false).expect("init BMP384");

    check_true!(bmp.osr_p == 4, "default_osr_p", passed, failed);
    check_true!(bmp.osr_t == 1, "default_osr_t", passed, failed);
    check_true!(bmp.iir   == 2, "default_iir",   passed, failed);

    let t = bmp.temperature().unwrap();
    check_true!(t >= -40.0 && t <= 85.0, "temperature_range", passed, failed);

    let p = bmp.pressure().unwrap();
    check_true!(p >= 300.0 && p <= 1250.0, "pressure_range", passed, failed);

    drop(bmp);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp_full = Bmp384Full::new(dev, addr, false).expect("init BMP384 Full");

    bmp_full.configure(2, 1, 1, 0x04).expect("configure");
    check_true!(bmp_full.inner.osr_p == 2 && bmp_full.inner.iir == 1 && bmp_full.inner.odr == 0x04,
        "configure_writes_through", passed, failed);

    bmp_full.set_mode(MODE_FORCED).expect("set_mode_forced");
    check_true!(bmp_full.inner.mode == MODE_FORCED, "set_mode_state", passed, failed);

    bmp_full.set_mode(MODE_NORMAL).expect("set_mode_normal");

    let ready = bmp_full.is_data_ready().unwrap_or(false);
    check_true!(ready == true || ready == false, "is_data_ready_runs", passed, failed);

    let alt = bmp_full.altitude(1013.25).unwrap_or(-1.0);
    check_true!(alt >= -500.0 && alt <= 9000.0, "altitude_range", passed, failed);

    let mut frames: [periph::chips::pressure::Bmp384FifoFrame; 16] = unsafe { std::mem::zeroed() };
    let _ = bmp_full.fifo_read(&mut frames).expect("fifo_read");

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}
